scio-core/src/main/scala/com/spotify/scio/values/SideInput.scala

```diff
@@ -22,19 +22,20 @@ import java.util.{List => JList, Map => JMap}
 
 import com.spotify.scio.util.JMapWrapper
 import org.apache.beam.sdk.transforms.DoFn
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow
 import org.apache.beam.sdk.values.PCollectionView
 
 import scala.collection.JavaConverters._
 
 /** Encapsulate an SCollection when it is being used as a side input. */
 trait SideInput[T] extends Serializable {
   private var cache: T = _
-  private var context: AnyRef = _
+  @transient private var window: BoundedWindow = _
```

Comment:
How about using

```
@transient lazy private val cache = Collections.synchronizedMap(new WeakHashMap[BoundedWindow, T]())
```

and in `getCache`:

```
cache.computeIfAbsent(window, _ => get(context))
```

reasoning:
 * all DoFn should share the same SideInput instance, this cache
 * there probably might be a case where one DoFn gets element from different window which would cause re-computation

One downside is that there is extra cost of synchronization on each retrieval, which could be mitigated by double locking on key, which would be sth like:

```
// note no synchronization on map
@transient lazy private val cache = new WeakHashMap[BoundedWindow, T]()
...
if (!cache.containsKey(window)) {
  synchronized(cache) {
    cache.computeIfAbsent(window, _ => get(context))
  }
}
cache.get(window)
```

Reply:
The other much bigger downside is GC of past windows. Unless if we make it a Guava LRU Cache or something instead of `WeakHashMap`. But that may require more tuning.
How about we keep it as is for now and optimize when there's a need?

The per-thread re-computation is trivial for most Java -> Scala collection mappers and probably only matters much for Sparkey/Annoy. And in those cases, the computed value lasts much longer (probably in batch?) and both Sparkey and Annoy share the underlying mmap.

Reply:
Is GC really a problem tho? Those objects need to be GC anyway, right? In case of `WeakHashMap` it would happen if needed. Could you benchmark this?